File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> lst = new LinkedList<>();
        inorder(root, lst);
        return lst;
    }

    private void inorder(TreeNode node, List<Integer> lst) {
        if (node == null) {
            return;
        }

        inorder(node.left, lst);
        lst.add(node.val);
        inorder(node.right, lst);
    }
}

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # def inorder(node, lst):
        #     if node == None:
        #         return lst
        #     inorder(node.left, lst)
        #     lst.append(node.val)
        #     inorder(node.right, lst)
        #     return lst
        # return inorder(root, [])
        def inorder(node):
            if not node:
                return []
            return inorder(node.left) + [node.val] + inorder(node.right)
        return inorder(root)

        # O(n) - space and time